source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -109,12 +109,15 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
              : Http::FilterHeadersStatus::Continue;
 }
 
-Http::FilterDataStatus Filter::decodeData(Buffer::Instance&, bool end_stream) {
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
   if (buffer_data_) {
     const bool buffer_is_full = isBufferFull();
     if (end_stream || buffer_is_full) {
       ENVOY_STREAM_LOG(debug, "ext_authz filter finished buffering the request since {}",
                        *callbacks_, buffer_is_full ? "buffer is full" : "stream is ended");
+      if (callbacks_->decodingBuffer() == nullptr && callbacks_->streamInfo().bytesReceived() > 0) {
+        callbacks_->addDecodedData(data, buffer_data_);
+      }
```

Comment:
I'm not sure this is correct in the `end_stream == true` case. If there are multiple calls to `decodeData` and the last has `end_stream` set, I think you still want to call `addDecodedData` if there is room in the buffer. Otherwise, during `initiateCall` (in `CheckRequestUtils::setHttpRequest`), the decoding buffer won't have the full request even though it would have fit in the buffer.

I think the logic should be:
```
if (buffer_data_) {
  const bool buffer_is_full = isBufferFull();
  if (end_stream || buffer_is_full) {
    // log...
    if (!buffer_is_full) {
      // Make sure data is available in initiateCall.
      callbacks_->addDecodedData(data, true);
    }
    initiateCall(*request_headers_);
    return filter_return_ == FilterReturn::StopDecoding
                     ? Http::FilterDataStatus::StopIterationAndWatermark
                     : Http::FilterDataStatus::Continue;
  } else {
    return Http::FilterDataStatus::StopIterationAndBuffer;
  }
}
```

There may be a way to organize that more cleanly.

Reply:
Hum, I agree with you. I'm not sure if we have a way to `setDecoderBufferLimit` in an integration test, so we can test that case here.